Add IEEE 802.15.4/6LoWPAN support

thvdveld opened the pull request:
This adds the IEEE 802.15.4 frame representation.
Still a work in progress and interested to know what you think about this.

I really would like to add 6LowPAN as well, however I'm not sure where to put this in smoltcp, since 6LowPAN is kind of weird.

thvdveld committed:
Add support for 802.15.4 and 6LoWPAN

Dirbaio commented on Cargo.toml:
```diff
@@ -35,12 +35,16 @@ alloc = ["managed/alloc"]
+"medium-sixlowpan" = ["socket", "sixlowpan", "proto-sixlowpan"]
+"phy-raw_socket" = ["std", "libc"]
 "phy-tuntap_interface" = ["std", "libc", "medium-ethernet"]
+"ieee802154" = []
```
features `ieee802154` and `sixlowpan` should probably be `proto-blahblah`

Dirbaio commented on src/dhcp/clientv4.rs:
```diff
@@ -0,0 +1,468 @@
+use super::{UDP_CLIENT_PORT, UDP_SERVER_PORT};
```
This file accidentally slipped through a rebase :)

thvdveld replied:
Oops :p Removed this!

Dirbaio commented on src/phy/mod.rs:
```diff
@@ -275,15 +277,36 @@ pub enum Medium {
     #[cfg(feature = "medium-ip")]
     Ip,
+
+    /// 6LoWPAN medium. Devices of this type send and receive IPv6 frames encoded using 6LoWPAN.
```
As implemented, the "Sixlowpan" medium is really the "Ieee802154" medium, because the Device impl sends/receives full Ieee802154 packets.

I would rename it (and the Cargo feature) to ieee802154. We could add a "Sixlowpan" medium in the future (not needed in this PR :smile: ) that really sends/receives only 6lowpan packets, without the ieee802154 header.

This would be consistent with the IP/Ethernet mediums:

- Medium::Ethernet: Ethernet header + IP packet
- Medium::Ip: IP packet
- Medium::Ieee802154: Ieee802154 header + 6lowpan packet
- Medium::Sixlowpan: 6lowpan packet

(6lowpan over BLE would require Medium::Sixlowpan, it doesn't have Ieee802154 headers)

thvdveld replied:
I renamed the feature to `medium-ieee802154` for now :smile:. I think that the features are more consistent now.

Dirbaio commented on src/wire/ieee802154.rs:
```diff
@@ -0,0 +1,931 @@
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Default)]
+pub struct Ieee802154Conf {
```
This struct is only used from Interface, perhaps it should be there instead of in `wire`. It looks like "internal runtime state", especially `sequence_no`.

thvdveld replied:
I removed it completely. Now the IEEE802.15.4 is stored in `hardware_addr` in the Interface. Not sure if an interface can have both an Ethernet address and a IEEE802.15.4 address, however, I can separate them in to to fields in `Interface` when needed. The source PAN ID and destination PAN ID are now stored in the context, since it is some kind of address I guess. But I'm not sure if it is the correct place to store them, since it is dynamic. Maybe it belongs in `Interface` as well.

Dirbaio commented on src/wire/sixlowpan.rs:
```diff
@@ -0,0 +1,1720 @@
+    Complete(ipv6::Address),
+    WithContext(&'a [u8]),
+    Elided,
+    Reserved,
```
Is this meant to represent "this bit combination is reserved, invalid for now"? If so it might make more sense to remove it and have src_addr/dst_addr return a Result, and represent this with Malformed.

In general it's nicer to represent "invalid" values with Result/Error, so the error is handled in the correct place, and no `unreachable!()` are required in places such as `resolve()`, which might cause crashes if someone forgets some check :)

thvdveld replied:
The `WithContext` is defined in the [RFC](https://datatracker.ietf.org/doc/html/rfc6282#section-3.1.1), however I don't know how to really use it.
Instead of returning `Malformed` I added the `NotSupported` error. This way, users will know that the behaviour is defined in some standard or RFC, but that smoltcp just does not support it yet. I think this can reduce debug time for users, since we already confirm that there is well-defined behaviour, just not supported.

Dirbaio reviewed (commented):
Thanks for your hard work! Just, wow, why is 6lowpan so complex with all the header compression and stuff? :rofl: 

I (finally, sorry for the slowness :sweat_smile:) got to do a first review. The structure and organization fits really well the existing code, so I see no major concerns there.

I've commented a few nits, mostly minor.

Something that needs fixing is the build fails when enabling `medium-sixlowpan` together with other mediums. This should be supported, to enable use cases like multi-medium devices (like an Ethernet + ieee802154 gateway). This is the reason for the `Medium` enum and all the `match`es on it.

Dirbaio committed:
wire: remove HardwareAddress::BROADCAST

Dirbaio committed:
ipv6: fix Solicited Node address calculation

Dirbaio committed:
6lowpan: do not fill neighbor cache from random packets

Equivalent of 6210612be047ee706ac729015cdbc2581e6ae9a3 for 6lowpan

Dirbaio committed:
ieee80154: process packets without the FCS.

We assume the FCS is checked by lower layers or by hardware.

- Makes it consistent with Ethernet mediums, where we don't handle the FCS either.
- Linux ieee802154 raw sockets send/receive packets without the FCS.

Dirbaio committed:
phy: Use right protocol on RawSocket based on the medium.

Dirbaio committed:
example/6lowpan: expand readme, do not use monitor interface.

Using a raw socket on `monitor0` causes weird results: packets we receive
include FCS, packets we send are parsed as if they didn't have FCS, except
by wireshark which always expects a FCS??

Turns out the sane way is to use raw sockets on normal `wpanX` interfaces,
in which case all packets we send/receive are without FCS.

Dirbaio committed:
Add RawHardwareAddress, use it in wire ndisc.

This avoids wire needing to know what medium we're on.

Dirbaio commented on src/iface/interface.rs:
```diff
@@ -989,6 +1101,103 @@ impl<'a> InterfaceInner<'a> {
+            SixlowpanNextHeader::Compressed => {
+                match SixlowpanNhcPacket::dispatch(payload)? {
+                    SixlowpanNhcPacket::ExtensionHeader(ext_header) => {
+                        todo!("{:?}", ext_header)
```
`todo!` makes smoltcp crash. Could you log some "this is unsupported" message and then ignore the packet instead?

Dirbaio commented on src/iface/interface.rs:
```diff
@@ -989,6 +1101,103 @@ impl<'a> InterfaceInner<'a> {
+                        // The packet wasn't handled by a socket, send an ICMP port unreachable packet.
+                        match ip_repr {
+                            #[cfg(feature = "proto-ipv6")]
+                            IpRepr::Ipv6(ipv6_repr) => {
```
ip_repr is always Sixlowpan, so this seems unreachable

Dirbaio commented on src/iface/interface.rs:
```diff
@@ -989,6 +1101,103 @@ impl<'a> InterfaceInner<'a> {
+                            }
+                            IpRepr::Unspecified { .. } => Err(Error::Unaddressable),
+                            IpRepr::Sixlowpan(_) => Err(Error::Malformed), // XXX(thvdveld): this is just wrong;
+                            r => todo!("{:?}", r),
```
same as above

Dirbaio commented on src/iface/interface.rs:
```diff
@@ -989,6 +1101,103 @@ impl<'a> InterfaceInner<'a> {
+                IpProtocol::Icmpv6 => {
+                    self.process_icmpv6(cx, sockets, ip_repr, iphc_packet.payload())
+                }
+                hdr => todo!("{:?}", hdr),
```
same as above

Dirbaio commented on src/wire/ip.rs:
```diff
@@ -500,6 +502,8 @@ pub enum Repr {
     #[cfg(feature = "proto-ipv6")]
     Ipv6(Ipv6Repr),
+    #[cfg(feature = "proto-sixlowpan")]
+    Sixlowpan(SixlowpanIphcRepr),
```
Adding `Sixlowpan` to `IpRepr` feels a bit strange to me. 6lowpan packets are just ipv6, but with a different on-wire format. It seems it should be possible to convert 6lowpan packets to/from the normal `IpRepr::Ipv6` when (de)compressing, so code that doesn't deal with 6lowpan doesn't have to care whether it's on 6lowpan or normal ipv6.

thvdveld replied:
Voila, removed `IpRepr::Sixlowpan`!

Dirbaio commented on src/iface/interface.rs:
```diff
@@ -229,32 +272,49 @@ let iface = InterfaceBuilder::new(device)
+                #[cfg(feature = "medium-ieee802154")]
+                sequence_no: self.sequence_no,
+                #[cfg(feature = "medium-ieee802154")]
+                src_pan_id: self.src_pan_id,
```
As discussed above, it'd be nice to have just a single `pan_id` instead of src/dst

Dirbaio reviewed (changes requested):
@thibautvdv I've been playing with the code, to learn more how it works, so here's the final "in depth" review :) 

The only "design issue" I see is adding the `IpRepr::Sixlowpan` variant, everything else is minor stuff.

thvdveld committed:
Use one PAN ID for source and destination

Also check for the correct destination PAN id when receiving a frame (as
discussed). Linux does this as well.
However, hardware implementations also can drop those packets.

thvdveld committed:
Use `net_debug` and drop instead of `todo!`

thvdveld committed:
Send ICMP unreachable when no UDP socket is found

thvdveld committed:
Set HardwareAddress behind feature flag

When there is no medium in the feature flags, then there is no
HardwareAddress.

thvdveld committed:
Defmt for RawHardwareAddress

thvdveld committed:
Remove some more todos

thvdveld committed:
Set flags to zero, because buffer could be reused.

thvdveld committed:
Remove IpRepr::Sixlowpan

thvdveld committed:
ieee802154: ignore frames with types other than Data

oddstr13 commented on src/iface/interface.rs:
```diff
@@ -1111,6 +1096,12 @@ impl<'a> InterfaceInner<'a> {
         let ieee802154_frame = Ieee802154Frame::new_checked(sixlowpan_payload)?;
         let ieee802154_repr = Ieee802154Repr::parse(&ieee802154_frame)?;
 
+        if ieee802154_repr.dst_pan_id != cx.pan_id {
```
This check is what stopped my code from working; I was not setting the PAN ID, so it defaults to `0x0000`.

This if is currently comparing `0x0000` with `None`, which presumably is not equal, and thus dropped.

Also not taking into consideration the `0xffff` broadcast PAN ID, which I don't know if is relevant to 6lowpan, but if it is, then the code should probably be capable of setting source and dest PAN ID separately, so it can broadcast and identify the source at the same time.
More research is needed about this.

thvdveld replied:
After some searching and looking for information in the IEEE802.15.4 standard it seems that a PAN id is required (so it cannot be `None` I guess). However, the PAN ID that is used by a network is selected by a PAN ID coordinator (a device that will assign a PAN ID and that handles PAN ID collisions etc.). This is not implemented here in `smoltcp` and I don't think it should be part of it. That should be some physical/mac crate for IEEE802.15.4.

I looked in the 6LoWPAN RFC but didn't find anything about PAN id's. So do we actually need to filter on the PAN id? It is not specified in the RFC. @Dirbaio looked in the Linux kernel and there frames are filtered, but I think those are filtered in `net/mac802154/rx.c` and thus not in the 6LoWPAN implementation of Linux (or I didn't look closely enough).

I think we shouldn't filter them? What do you guys think? We could write a note in the README whenever we merge this into master to tell them that we accept all PAN id's (however not all hardware IDs).

For now I pushed a commit where we filter on our own PAN id AND broadcast.

Dirbaio replied:
I'm pretty sure the right behavior is to filter on pan_id. This is what Linux does. It makes sense, since it allows having multiple independent PANs where a broadcast on one isn't received on the others for example, just like independent WiFi networks sharing the air.

thvdveld committed:
ieee802154: Correct filtering of PAN id

thvdveld committed:
ieee802154: update documentation

thvdveld committed:
ieee802154: log when we drop a frame

Dirbaio committed:
Fix some comments.

Dirbaio committed:
Add medium-ieee802154 to CI.

Not adding to `defmt` because it doesn't build yet.

Dirbaio committed:
Fix build when setting medium-ieee802154 but not proto-sixlowpan.

Dirbaio reviewed (approved):
@thibautvdv I'm going to merge this. It's complete, works great, code is well structured! :+1: The few remaining TODOs don't seem very critical, they can be addressed in future PRs.

I want to do some big changes to Interface code, which would send this to conflict hell otherwise.
